File: python/math/50. Pow(x, n).py
'''
Recursion
Time: O(log n)
Space: O(log n)

n is even: x^n = (x^2)^(n/2)
n is odd: x^n = x * (x^2)^((n-1)/2)
'''

class Solution:
    def myPow(self, x: float, n: int) -> float:
        def helper(x, n):
            if x == 0: return 0
            if n == 0: return 1

            res = helper(x * x, n // 2)
            return res * x if n % 2 else res
        
        res = helper(x, abs(n))
        return res if n >=0 else 1 / res

File: javascript/math/50. Pow(x, n).js
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
    const helper = (x, n) => {
        if (x === 0) return 0;
        if (n === 0) return 1;

        res = helper(x * x, Math.floor(n / 2));
        return n % 2 ? res * x : res;
    }
    res = helper(x, Math.abs(n));
    return n >= 0 ? res : 1 / res;
};

var myPow = function (x, n) {
    if (n == 0) return 1;
    if (x == 0) return 0;
    if (n < 0) {
      n = -n;
      x = 1 / x;
    }

    let res = 1;
    while(n) {
      if (n & 1) res *= x;
      x *= x;
      n >>>= 1;
    }
    return res;
  };